
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Functions &#8212; Bufalo 0.01 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="As a R Script" href="Usage.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="functions">
<h1>Functions<a class="headerlink" href="#functions" title="Permalink to this headline">¶</a></h1>
<div class="section" id="import-data-from-ballgown-object-code-99">
<h2>Import Data From Ballgown Object (code = 99)<a class="headerlink" href="#import-data-from-ballgown-object-code-99" title="Permalink to this headline">¶</a></h2>
<p>This function allows to create all the proper files to be used in the
other functions taking advantage of the <a class="reference external" href="http://bioconductor.org/packages/release/bioc/html/ballgown.html">Ballgown</a>
package (ballgown function). To gain a better understanding of this
function, check the Ballgown documentation
(<a class="reference external" href="https://bioconductor.org/packages/release/bioc/manuals/ballgown/man/ballgown.pdf">https://bioconductor.org/packages/release/bioc/manuals/ballgown/man/ballgown.pdf</a>)</p>
<div class="section" id="syntax">
<h3>Syntax:<a class="headerlink" href="#syntax" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Rscript</span> <span class="n">reader</span><span class="o">.</span><span class="n">R</span> <span class="mi">99</span> <span class="o">&lt;</span><span class="n">Mapping</span> <span class="n">file</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">Folder</span> <span class="n">Containing</span> <span class="n">The</span> <span class="n">Data</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">Common</span> <span class="n">part</span> <span class="n">of</span> <span class="n">samples</span> <span class="n">name</span><span class="o">&gt;</span>
<span class="n">python</span> <span class="n">rpy_reader</span><span class="o">.</span><span class="n">py</span> <span class="mi">99</span> <span class="o">&lt;</span><span class="n">Mapping</span> <span class="n">file</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">Folder</span> <span class="n">Containing</span> <span class="n">The</span> <span class="n">Data</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">Common</span> <span class="n">part</span> <span class="n">of</span> <span class="n">samples</span> <span class="n">name</span><span class="o">&gt;</span>
</pre></div>
</div>
<p>Arguments:</p>
<ul>
<li><p class="first">Mapping File: data frame with rows corresponding to samples and
columns corresponding to phenotipic variables. Up to now, this file
was structured as follows:</p>
<table border="1" class="docutils">
<colgroup>
<col width="39%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">ids</th>
<th class="head">tissue</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>id 1</td>
<td>name 1</td>
</tr>
<tr class="row-odd"><td>id 2</td>
<td>name 2</td>
</tr>
</tbody>
</table>
<p>Reporting only the tissue name. If a file presents more columns,
there should be no problem because they will be ignored. This file
will be used as the argument pData in the “ballgown” function
(<a class="reference external" href="https://bioconductor.org/packages/release/bioc/manuals/ballgown/man/ballgown.pdf">Ballgown documentation</a>).</p>
</li>
<li><p class="first">Folder Containing The Data: folder containing the results from the
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/27560171">pipeline</a> described
by Petrea and collegues.</p>
</li>
<li><p class="first">Common part of samples name: a common root or part of samples name
needed by ballgown to extract properl all the data needed.</p>
</li>
</ul>
<p><strong>This function is not available for the shiny app: this step must be executed via command line</strong></p>
</div>
</div>
<div class="section" id="plotter-code-1">
<h2>Plotter (code=1)<a class="headerlink" href="#plotter-code-1" title="Permalink to this headline">¶</a></h2>
<p>Plots the FPKM values of all transcripts of a given gene. This
function takes advantage of the “plotTranscript” function in the
ballgown package. For further informations, check the <a class="reference external" href="https://bioconductor.org/packages/release/bioc/manuals/ballgown/man/ballgown.pdf">Ballgown
documentation</a></p>
<div class="section" id="id2">
<h3>Syntax:<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Rscript</span> <span class="n">reader</span><span class="o">.</span><span class="n">R</span> <span class="mi">1</span> <span class="o">&lt;</span><span class="n">gene</span><span class="o">&gt;</span>
<span class="n">python3</span> <span class="n">rpy_reader</span><span class="o">.</span><span class="n">py</span> <span class="mi">1</span> <span class="o">&lt;</span><span class="n">gene</span><span class="o">&gt;</span>
</pre></div>
</div>
<ul class="simple">
<li>gene: the gene symbol (name) to be analyzed.</li>
</ul>
</div>
</div>
<div class="section" id="search-by-tissue-code-2">
<h2>Search By Tissue (code=2)<a class="headerlink" href="#search-by-tissue-code-2" title="Permalink to this headline">¶</a></h2>
<p>Searches the a given tissue and writes a file containing the FPKM
vaules of all genes. If a second argument with the name of a gene is
given, the FPKM of this gene insider the chosen Tissue.</p>
<div class="section" id="id3">
<h3>Syntax:<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Rscript</span> <span class="n">reader</span><span class="o">.</span><span class="n">R</span> <span class="mi">2</span> <span class="o">&lt;</span><span class="n">tissue</span><span class="o">&gt;</span> <span class="p">[</span><span class="n">gene</span><span class="p">]</span>
<span class="n">python3</span> <span class="n">rpy_reader</span><span class="o">.</span><span class="n">py</span> <span class="mi">1</span> <span class="o">&lt;</span><span class="n">tissue</span><span class="o">&gt;</span> <span class="p">[</span><span class="n">gene</span><span class="p">]</span>
</pre></div>
</div>
<ul class="simple">
<li>tissue: the tissue name to be analyzed.</li>
<li>gene: [optional] the gene symbol (name) to be analyzed.</li>
</ul>
</div>
</div>
<div class="section" id="search-by-gene-code-3">
<h2>Search By Gene (code=3)<a class="headerlink" href="#search-by-gene-code-3" title="Permalink to this headline">¶</a></h2>
<p>Searches the FPKM values of a single gene in all tissues.</p>
<div class="section" id="id4">
<h3>Syntax:<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Rscript</span> <span class="n">reader</span><span class="o">.</span><span class="n">R</span> <span class="mi">3</span> <span class="o">&lt;</span><span class="n">gene</span><span class="o">&gt;</span>
<span class="n">python3</span> <span class="n">rpy_reader</span><span class="o">.</span><span class="n">py</span> <span class="mi">3</span> <span class="o">&lt;</span><span class="n">gene</span><span class="o">&gt;</span>
</pre></div>
</div>
<ul class="simple">
<li>gene: the gene symbol (name) to be analyzed.</li>
</ul>
</div>
</div>
<div class="section" id="search-by-gene-feature-code-4">
<h2>Search By Gene Feature (code=4)<a class="headerlink" href="#search-by-gene-feature-code-4" title="Permalink to this headline">¶</a></h2>
<p>Allows to analyze the expression of gene feature (exons and introns).</p>
<div class="section" id="id5">
<h3>Syntax:<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Rscript</span> <span class="n">reader</span><span class="o">.</span><span class="n">R</span> <span class="mi">4</span> <span class="o">&lt;</span><span class="n">gene</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">feature</span><span class="o">&gt;</span>
<span class="n">python3</span> <span class="n">rpy_reader</span><span class="o">.</span><span class="n">py</span> <span class="mi">4</span> <span class="o">&lt;</span><span class="n">gene</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">feature</span><span class="o">&gt;</span>
</pre></div>
</div>
<ul class="simple">
<li>gene: the gene symbol (name) to be analyzed.</li>
<li>feature: the gene feature to analyzed. Possible values: intron, exon.</li>
</ul>
</div>
</div>
<div class="section" id="search-by-differential-fold-code-5">
<h2>Search By Differential Fold (code=5)<a class="headerlink" href="#search-by-differential-fold-code-5" title="Permalink to this headline">¶</a></h2>
<p>Given 2 tissues, the differential fold (linear and log2) is
calculated. To avoid “Inf” values, to all FPKM values was added
0.000001. If a fourth argument is given reporting a gene name, the
function will consider only the entries containing only that
gene.</p>
<div class="section" id="id6">
<h3>Syntax:<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Rscript</span> <span class="n">reader</span><span class="o">.</span><span class="n">R</span> <span class="mi">5</span> <span class="o">&lt;</span><span class="n">tissue</span> <span class="mi">1</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">tissue</span> <span class="mi">2</span><span class="o">&gt;</span> <span class="p">[</span><span class="n">gene</span><span class="p">]</span>
<span class="n">python3</span> <span class="n">rpy_reader</span><span class="o">.</span><span class="n">py</span> <span class="mi">5</span> <span class="o">&lt;</span><span class="n">tissue</span> <span class="mi">1</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">tissue</span> <span class="mi">2</span><span class="o">&gt;</span> <span class="p">[</span><span class="n">gene</span><span class="p">]</span>
</pre></div>
</div>
<ul class="simple">
<li>tissue 1: First tissue to be considered.</li>
<li>tissue 2: Second tissue to be considered.</li>
<li>gene: [optional] the gene symbol (name) to be analyzed.</li>
</ul>
</div>
</div>
<div class="section" id="search-by-differential-fold-code-6">
<h2>Search By Differential Fold (code=6)<a class="headerlink" href="#search-by-differential-fold-code-6" title="Permalink to this headline">¶</a></h2>
<p>Given a gene symbol, the coexpression module, found with WGCNA is
returned.</p>
<div class="section" id="id7">
<h3>Syntax:<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Rscript</span> <span class="n">reader</span><span class="o">.</span><span class="n">R</span> <span class="mi">6</span> <span class="o">&lt;</span><span class="n">gene</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">ModuleFile</span><span class="o">&gt;</span>
<span class="n">python3</span> <span class="n">rpy_reader</span><span class="o">.</span><span class="n">py</span> <span class="mi">6</span> <span class="o">&lt;</span><span class="n">gene</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">ModuleFile</span><span class="o">&gt;</span>
</pre></div>
</div>
<ul>
<li><p class="first">gene: the gene symbol (name) to be analyzed.</p>
</li>
<li><p class="first">File containing the coexpression module found with <a class="reference external" href="https://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/">WGCNA</a>.</p>
<p><strong>AGGIUNGERE LO SCRIPT PER LA COESPRESSIONE NELLA REPO</strong></p>
</li>
</ul>
</div>
</div>
<div class="section" id="id8">
<h2>Search By Differential Fold (code=6)<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h2>
<p>Given a gene symbol, from the coexpression module, found with <a class="reference external" href="https://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/">WGCNA</a>
a network graph containing only n nodes (genes) over a certain
correlation value is produced Arguments:</p>
<div class="section" id="id10">
<h3>Syntax:<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Rscript</span> <span class="n">reader</span><span class="o">.</span><span class="n">R</span> <span class="mi">7</span> <span class="o">&lt;</span><span class="n">symbol</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">gene</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">ModuleFile</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">corr</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">nodes</span><span class="o">&gt;</span>
<span class="n">python3</span> <span class="n">rpy_reader</span><span class="o">.</span><span class="n">py</span> <span class="mi">7</span> <span class="o">&lt;</span><span class="n">symbol</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">gene</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">ModuleFile</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">corr</span><span class="o">&gt;</span> <span class="o">&lt;</span><span class="n">nodes</span><span class="o">&gt;</span>
</pre></div>
</div>
<ul>
<li><p class="first">symbol: a legacy argument. Use symbol as first argument (to removed
in the next versions).</p>
</li>
<li><p class="first">gene: the gene symbol (name) to be analyzed.</p>
</li>
<li><p class="first">File containing the coexpression module found with <a class="reference external" href="https://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/Rpackages/WGCNA/">WGCNA</a>.</p>
</li>
<li><p class="first">Correlation threashold value to show a connection between 2 genes.</p>
</li>
<li><p class="first">Number of nodes-genes to be shown.</p>
<p><strong>AGGIUNGERE LO SCRIPT PER LA COESPRESSIONE NELLA REPO</strong></p>
</li>
</ul>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Functions</a><ul>
<li><a class="reference internal" href="#import-data-from-ballgown-object-code-99">Import Data From Ballgown Object (code = 99)</a><ul>
<li><a class="reference internal" href="#syntax">Syntax:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#plotter-code-1">Plotter (code=1)</a><ul>
<li><a class="reference internal" href="#id2">Syntax:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#search-by-tissue-code-2">Search By Tissue (code=2)</a><ul>
<li><a class="reference internal" href="#id3">Syntax:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#search-by-gene-code-3">Search By Gene (code=3)</a><ul>
<li><a class="reference internal" href="#id4">Syntax:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#search-by-gene-feature-code-4">Search By Gene Feature (code=4)</a><ul>
<li><a class="reference internal" href="#id5">Syntax:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#search-by-differential-fold-code-5">Search By Differential Fold (code=5)</a><ul>
<li><a class="reference internal" href="#id6">Syntax:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#search-by-differential-fold-code-6">Search By Differential Fold (code=6)</a><ul>
<li><a class="reference internal" href="#id7">Syntax:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id8">Search By Differential Fold (code=6)</a><ul>
<li><a class="reference internal" href="#id10">Syntax:</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="Usage.html" title="previous chapter">As a R Script</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/Functions.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, Ilario Tagliaferri.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.7.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/Functions.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>